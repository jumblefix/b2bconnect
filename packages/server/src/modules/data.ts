export const bookData = {
  title: 'My Book',
  coverImage: 'my-book.jpg',
  isbn: '1233434234',
  description:
    'Right now when I run npm test the output is just: ' +
    'and the number of passed and failed tests (when the tests are successful). ' +
    'I would like the output to include  and any other test descriptions' +
    ' (like how the output looks when an test is run).',
  rating: 0,
  price: 66.0,
  offerPrice: 56.0,
  categoryId: 'invalid',
  publishedYear: 1999,
};

export const user = {
  name: 'example',
  email: 'example@email.com',
  password: '123456',
};
