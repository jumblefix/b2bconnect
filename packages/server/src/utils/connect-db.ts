// tslint:disable-next-line:no-var-requires
const iconv = require('iconv-lite');
// tslint:disable-next-line:no-var-requires
const encodings = require('iconv-lite/encodings');
iconv.encodings = encodings;

import { createConnection, getConnectionOptions } from 'typeorm';

export const connectDb = async () => {
  const connectionOptions = await getConnectionOptions(process.env.NODE_ENV);

  return createConnection({
    ...connectionOptions,
    name: 'default',
  });
};

export const connectDbTest = async (resetDB: boolean = false) => {
  const connectionOptions = await getConnectionOptions(process.env.NODE_ENV);

  return createConnection({
    ...connectionOptions,
    name: 'default',
    synchronize: resetDB,
    dropSchema: resetDB,
  });
};
