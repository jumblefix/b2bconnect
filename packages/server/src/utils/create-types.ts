import { generateNamespace } from '@gql2ts/from-schema';
import * as fs from 'fs';
import * as path from 'path';
import { genSchema } from './schema-utils';

const typescriptTypes = generateNamespace('GQL', genSchema());

fs.writeFile(
  path.join(__dirname, '../typings/GQL/index.d.ts'),
  typescriptTypes,
  err => {
    console.error(err);
  },
);
