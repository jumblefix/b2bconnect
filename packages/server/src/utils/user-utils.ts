import { Request, Response } from 'express';
import * as IORedis from 'ioredis';
import { v4 } from 'uuid';
import { ValidationError } from 'yup';
import {
  FORGOT_PASSWORD_PREFIX,
  REDIS_SESSION_PREFIX,
  TokenTypes,
  USER_SESSION_PREFIX,
} from '../constants';
import { User } from '../entity/User';
import { GraphQLMiddlewareFunc, Resolver } from '../typings/app-utility-types';

export const createForgotPasswordLink = async (
  url: string,
  userId: string,
  redis: IORedis.Redis,
) => {
  const id = v4();
  await redis.set(`${FORGOT_PASSWORD_PREFIX}${id}`, userId, 'ex', 60 * 60);
  return `${url}/change-password/${id}`;
};

export const removeAllUsersSessions = async (
  userId: string,
  redis: IORedis.Redis,
) => {
  const sessionIds = await redis.lrange(
    `${USER_SESSION_PREFIX}${userId}`,
    0,
    -1,
  );

  const promises = [];
  // tslint:disable-next-line:prefer-for-of
  for (let i = 0; i < sessionIds.length; i += 1) {
    promises.push(redis.del(`${REDIS_SESSION_PREFIX}${sessionIds[i]}`));
  }
  await Promise.all(promises);
};

export const formatYupError = (err: ValidationError) => {
  const errors: [{path: string; message: string}] = [] as any;
  err.inner.forEach(e => {
    errors.push({
      path: e.path,
      message: e.message,
    });
  });

  return errors;
};

export const createMiddleware = (
  middlewareFunc: GraphQLMiddlewareFunc,
  resolverFunc: Resolver,
) => (parent: any, args: any, context: any, info: any) =>
  middlewareFunc(resolverFunc, parent, args, context, info);

export const middleware = async (
  resolver: Resolver,
  parent: any,
  args: any,
  context: any,
  info: any,
) => {
  return resolver(parent, args, context, info);
};

export const createTokenLink = async (
  url: string,
  userId: string,
  redis: IORedis.Redis,
  type: TokenTypes,
) => {
  const id = v4();
  await redis.set(id, userId, 'ex', 60 * 60 * 24);
  return `${url}/${type}/${id}`;
};

export const confirmEmail = async (req: Request, res: Response) => {
  const { id } = req.params;
  const RedisInstance = new IORedis();
  const userId = await RedisInstance.get(id);
  if (userId) {
    await User.update({ id: userId }, { confirmed: true });
    await RedisInstance.del(id);
    res.redirect(`${process.env.FRONTEND_HOST}/login`);
  } else {
    res.json({ message: 'invalid or expired register token' });
  }
};

export const resetPassword = async (req: Request, res: Response) => {
  const { id } = req.params;
  const RedisInstance = new IORedis();
  const userId = await RedisInstance.get(id);
  if (userId) {
    res.redirect(`${process.env.FRONTEND_HOST}/reset/${id}`);
  } else {
    res.json({ message: 'invalid or expired reset token' });
  }
};
