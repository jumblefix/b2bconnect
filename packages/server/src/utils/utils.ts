import {
  formatError as formatApolloError,
  isInstance as isApolloErrorInstance,
} from 'apollo-errors';
import * as nodemailer from 'nodemailer';
import slugify from 'slugify';
import { ValidationError } from 'yup';
import { Env } from '../constants';
import {
  AuthenticationError,
  AuthorizationError,
  InputValidationError,
} from './errors';

export const add = (...args: number[]) => args.reduce((c, p) => c + p);

export const formatYupError = (err: ValidationError) => {
  const errors: [{ path: string; message: string }] = [] as any;
  err.inner.forEach(e => {
    errors.push({
      path: e.path,
      message: e.message,
    });
  });
  return errors;
};

export function formatError(error: any) {
  const { originalError } = error;
  if (isApolloErrorInstance(originalError)) {
    // log internalData to stdout but not include it in the formattedError
    console.error(
      JSON.stringify({
        type: `error`,
        data: originalError.data,
        internalData: originalError.internalData,
      }),
    );
  }
  return formatApolloError(error);
}

export const makeSlug = (str: string) => slugify(str, { lower: true });

export const sendEmail = async ({
  to,
  subject,
  text,
  html,
}: SendEmailParams) => {
  let mailConfig: any;
  if (process.env.NODE_ENV === Env.production) {
    mailConfig = {
      host: 'smtp.sendgrid.net',
      port: 587,
      auth: {
        user: 'real.user',
        pass: 'verysecret',
      },
    };
  } else {
    mailConfig = {
      host: process.env.MAIL_HOST,
      port: process.env.MAIL_PORT,
      auth: {
        user: process.env.MAIL_USER,
        pass: process.env.MAIL_PASS,
      },
    };
  }

  const transport = nodemailer.createTransport(mailConfig);

  try {
    const info = await transport.sendMail({
      to,
      subject,
      text,
      html,
    });

    console.log(nodemailer.getTestMessageUrl(info));
  } catch (error) {
    console.error(error);
    throw new Error(error);
  }
  return true;
};

export const checkAdminRights = (session: any) => {
  const { userId, isAdmin } = session;
  if (!userId) {
    throw new AuthenticationError();
  }
  if (!isAdmin) {
    throw new AuthorizationError();
  }
};

export const validateInputs = async (schema: any, inputs: any) => {
  try {
    await schema.validate(inputs, { abortEarly: false });
  } catch (err) {
    const errors = formatYupError(err);
    throw new InputValidationError({
      data: errors,
    });
  }
};
