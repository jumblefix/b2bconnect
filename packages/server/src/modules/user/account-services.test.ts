import { graphql } from 'graphql';
import * as ioredis from 'ioredis';
import { Connection } from 'typeorm';
import { connectDbTest } from '../../utils/connect-db';
import { createDb } from '../../utils/create-db';
import { genSchema } from '../../utils/schema-utils';
import { user } from '../data';

let connection: Connection;
let redis: ioredis.Redis;
let registerId: string;
let registeredEmail: string;

beforeAll(async () => {
  await createDb();
  redis = new ioredis();
  connection = await connectDbTest(true);
});

afterAll(async () => {
  connection.close();
});

const resendVerifySignup = {
  caseId: 'resendVerifySignup',
  query: `mutation{ resendVerifySignup }`,
  session: { userId: '123' },
  expectation: (result: any) => {
    expect(result).toMatchSnapshot();
  },
};

const resendVerifySignupWoLogin = {
  caseId: 'resendVerifySignup',
  query: `mutation{ resendVerifySignup }`,
  session: { userId: '' },
  expectation: (result: any) => {
    expect(result).toMatchSnapshot();
  },
};

const sendResetPasswordInvalidEmail = {
  caseId: 'sendResetPasswordInvalidEmail',
  query: `mutation{ sendResetPassword(email:"invalid") }`,
  session: { userId: '' },
  expectation: (result: any) => {
    expect(result).toMatchSnapshot();
  },
};

const registerUserValidData = {
  caseId: 'registerUserValidData',
  query: `mutation {
               register(name:"${user.name}", email: "${
    user.email
  }", password: "${user.password}") { email name id }}`,
  session: { userId: '' },
  expectation: (result: any) => {
    registerId = result.data!.register.id;
    registeredEmail = result.data!.register.email;
    expect(result.data!.register.email).toEqual(user.email);
  },
};

const sendResetPassword = (email: string, id: string) => {
  return {
    caseId: 'sendResetPassword',
    query: `mutation{ sendResetPassword(email:"${email}") }`,
    session: { userId: id },
    expectation: (result: any) => {
      expect(result).toMatchSnapshot();
    },
  };
};

const changePassword = (_: string, id: string) => {
  return {
    caseId: 'changePassword',
    query: `mutation{ changePassword(oldPassword:"123456",password:"1234567"){id} }`,
    session: { userId: id },
    expectation: (result: any) => {
      expect(result).toEqual({ data: { changePassword: { id } } });
    },
  };
};

const changePasswordInvalidOldPassword = (_: string, id: string) => {
  return {
    caseId: 'changePassword',
    query: `mutation{ changePassword(oldPassword:"invalid",password:"1234567"){id} }`,
    session: { userId: id },
    expectation: (result: any) => {
      expect(result).toMatchSnapshot();
    },
  };
};

const changePasswordWoLogin = {
  caseId: 'changePassword',
  query: `mutation{ changePassword(oldPassword:"123456",password:"1234567"){id} }`,
  session: { userId: '' },
  expectation: (result: any) => {
    expect(result).toMatchSnapshot();
  },
};

const changePasswordInvalidUser = {
  ...changePasswordWoLogin,
  caseId: 'changePasswordInvalidUser',
  session: { userId: '123' },
  expectation: (result: any) => {
    expect(result).toMatchSnapshot();
  },
};

const changeEmail = (_: string, id: string) => {
  return {
    caseId: 'changeEmail',
    query: `mutation{ changeEmail(email:"new@email.com"){email} }`,
    session: { userId: id },
    expectation: (result: any) => {
      expect(result).toMatchSnapshot();
    },
  };
};

const changeEmailSame = (_: string, id: string) => {
  return {
    caseId: 'changeEmail',
    query: `mutation{ changeEmail(email:"new@email.com"){email} }`,
    session: { userId: id },
    expectation: (result: any) => {
      expect(result).toMatchSnapshot();
    },
  };
};

const changeEmailWoLogin = {
  caseId: 'changeEmail',
  query: `mutation{ changeEmail(email:"new@email.com"){email} }`,
  session: { userId: '' },
  expectation: (result: any) => {
    expect(result).toMatchSnapshot();
  },
};

const changeEmailInvalidUser = {
  ...changeEmailWoLogin,
  caseId: 'changeEmailInvalidUser',
  session: { userId: '123' },
  expectation: (result: any) => {
    expect(result).toMatchSnapshot();
  },
};
const verifyResetPassword = {
  query: `mutation{
    verifyResetPassword
    (token:"123",password:"password1",confirmPassword:"password2",)
    {email}}`,
  caseId: 'verifyResetPassword',
  session: { userId: '123' },
  expectation: (result: any) => {
    expect(result).toMatchSnapshot();
  },
};

describe('account services', () => {
  const cases = [
    resendVerifySignup,
    resendVerifySignupWoLogin,
    sendResetPasswordInvalidEmail,
    registerUserValidData,
    sendResetPassword,
    changePassword,
    changePasswordInvalidOldPassword,
    changePasswordWoLogin,
    changePasswordInvalidUser,
    changeEmail,
    changeEmailSame,
    changeEmailWoLogin,
    changeEmailInvalidUser,
    verifyResetPassword,
  ];

  cases.forEach(c => {
    it(`case:`, async () => {
      const { query, expectation, session } =
        typeof c === 'function' ? c(registeredEmail, registerId) : c;

      const ctx = {
        redis,
        session,
        db: connection,
        url: 'http://localhost:4000',
        req: {},
      };

      const result = await graphql(genSchema(), query, null, ctx, {});

      expectation(result);
    });
  });
});
