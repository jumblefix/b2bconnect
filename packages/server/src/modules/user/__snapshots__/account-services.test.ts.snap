// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`account services case: 1`] = `
Object {
  "data": Object {
    "resendVerifySignup": true,
  },
}
`;

exports[`account services case: 2`] = `
Object {
  "data": Object {
    "resendVerifySignup": null,
  },
  "errors": Array [
    [GraphQLError: Please login to continue],
  ],
}
`;

exports[`account services case: 3`] = `
Object {
  "data": Object {
    "sendResetPassword": null,
  },
  "errors": Array [
    [GraphQLError: User not found],
  ],
}
`;

exports[`account services case: 4`] = `
Object {
  "data": Object {
    "sendResetPassword": true,
  },
}
`;

exports[`account services case: 5`] = `
Object {
  "data": Object {
    "changePassword": null,
  },
  "errors": Array [
    [GraphQLError: Invalid old password],
  ],
}
`;

exports[`account services case: 6`] = `
Object {
  "data": Object {
    "changePassword": null,
  },
  "errors": Array [
    [GraphQLError: Please login to continue],
  ],
}
`;

exports[`account services case: 7`] = `
Object {
  "data": Object {
    "changePassword": null,
  },
  "errors": Array [
    [GraphQLError: User not found],
  ],
}
`;

exports[`account services case: 8`] = `
Object {
  "data": Object {
    "changeEmail": Object {
      "email": "new@email.com",
    },
  },
}
`;

exports[`account services case: 9`] = `
Object {
  "data": Object {
    "changeEmail": null,
  },
  "errors": Array [
    [GraphQLError: New email is same as old one],
  ],
}
`;

exports[`account services case: 10`] = `
Object {
  "data": Object {
    "changeEmail": null,
  },
  "errors": Array [
    [GraphQLError: Please login to continue],
  ],
}
`;

exports[`account services case: 11`] = `
Object {
  "data": Object {
    "changeEmail": null,
  },
  "errors": Array [
    [GraphQLError: User not found],
  ],
}
`;

exports[`account services case: 12`] = `
Object {
  "data": Object {
    "verifyResetPassword": null,
  },
  "errors": Array [
    [GraphQLError: Passwords do not match],
  ],
}
`;
